lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,280 @@
+	Mode        types.SessionParticipantMode
+}
+
+// NewSessionStream creates a new session stream.
+// The type of the handshake parameter determines if this is the client or server end.
+func NewSessionStream(conn *websocket.Conn, handshake interface{}) (*SessionStream, error) {
+	s := &SessionStream{
+		conn:           conn,
+		in:             make(chan []byte),
+		done:           make(chan struct{}),
+		resizeQueue:    make(chan *remotecommand.TerminalSize, 1),
+		forceTerminate: make(chan struct{}),
+	}
+
+	clientHandshake, isClient := handshake.(ClientHandshake)
+	serverHandshake, ok := handshake.(ServerHandshake)
+
+	if !isClient && !ok {
+		return nil, trace.BadParameter("Handshake must be either client or server handshake, got %T", handshake)
+	}
+
+	if isClient {
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("Expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ServerHandshake == nil {
+			return nil, trace.Errorf("Expected websocket server handshake, got %v", msg)
+		}
+
+		s.MFARequired = msg.ServerHandshake.MFARequired
+		handshakeMsg := metaMessage{ClientHandshake: &clientHandshake}
+		dataClientHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataClientHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		handshakeMsg := metaMessage{ServerHandshake: &serverHandshake}
+		dataServerHandshake, err := utils.FastMarshal(handshakeMsg)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := conn.WriteMessage(websocket.TextMessage, dataServerHandshake); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		ty, data, err := conn.ReadMessage()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if ty != websocket.TextMessage {
+			return nil, trace.Errorf("Expected websocket control message, got %v", ty)
+		}
+
+		var msg metaMessage
+		if err := utils.FastUnmarshal(data, &msg); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if msg.ClientHandshake == nil {
+			return nil, trace.Errorf("Expected websocket client handshake")
+		}
+
+		s.Mode = msg.ClientHandshake.Mode
+	}
+
+	go s.readTask()
+	return s, nil
+}
+
+func (s *SessionStream) readTask() {
+	for {
+		terminated := false
+		defer s.closedC.Do(func() { close(s.done) })
+		ty, data, err := s.conn.ReadMessage()
+		if err != nil {
+			return
+		}
+
+		if ty == websocket.BinaryMessage {
+			s.in <- data
+		}
+
+		if ty == websocket.TextMessage {
+			var msg metaMessage
+			if err := utils.FastUnmarshal(data, &msg); err != nil {
+				return
+			}
+
+			if msg.Resize != nil {
+				s.resizeQueue <- msg.Resize
+			}
+
+			if msg.ForceTerminate && !terminated {
+				terminated = true
```

Comment:
Mentioned this above, but this write will be overwritten to false on the next iteration of the loop, before you get a chance to read it.